src/transaction/log_page_buffer.c

```diff
@@ -1988,6 +1995,31 @@ logpb_copy_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_CS_ACCESS_MODE
   return rv;
 }
 
+static void
+request_log_page_from_ps (LOG_PAGEID log_pageid)
+{
+#if defined(SERVER_MODE)
+  // *INDENT-OFF*
+  constexpr size_t BIG_INT_SIZE = 8;
+  char buffer[BIG_INT_SIZE];
+  std::memcpy (buffer, &log_pageid, sizeof (log_pageid));
+  std::string message (buffer, BIG_INT_SIZE);
+
+  if (!ats_Gl.get_log_page_receiver ().is_page_requested (log_pageid))
+    {
+      ats_Gl.push_request (ats_to_ps_request::SEND_LOG_PAGE_FETCH, std::move (message));
+
+      if (prm_get_bool_value (PRM_ID_ER_LOG_READ_LOG_PAGE))
+        {
+          _er_log_debug (ARG_FILE_LINE, "Sent request for log to Page Server. Page ID: %lld \n", log_pageid);
+        }
+    }
+  
+  ats_Gl.get_log_page_receiver ().set_page_requested (log_pageid);
```

Comment:
This is not entirely safe because it allows the possibility of multiple requests for the same log_pageid. Take the next scenario with thread T1 and T2, making a request for log page LP.

- receiver has no request for LP initially
- T1 checks is_page_requested (LP) and gets false.
- T2 checks is_page_requested (LP) and gets false.
- T1 & T2 push separate requests for same LP
- T1 & T2 register a single entry in log page receiver, with two requesters.
- Receiver will get two copies of LP from the page server, because two requests are sent. After receiving the first copy, it serves them to both T1 & T2 and erases the map entries for LP.
- The second copy is received, but there is no requester left to serve.

There are some possible fixes for this case (e.g. discard the second copy).

Reply:
I will ignore the log pages received from PS if there's no request ongoing for them (IE the requests have been fulfilled).